components/locale_core/src/preferences/locale.rs

```diff
@@ -27,31 +27,59 @@ pub struct LocalePreferences {
 }
 
 impl LocalePreferences {
-    fn to_data_locale_maybe_region_priority(self, region_priority: bool) -> DataLocale {
+    /// Convert to a DataLocale, with region-based fallback priority
+    ///
+    /// Most users should use `icu_provider::marker::make_locale()` instead.
+    pub fn to_data_locale_region_priority(self) -> DataLocale {
+        let (region, subdivision) = if let Some(ro) = self.region_override {
+            // Use the region override if present
+            (
+                Some(ro.region),
+                Some(*ro).filter(|sd| !sd.suffix.is_unknown()),
+            )
+        } else if let Some(sd) = self.subdivision {
+            if let Some(r) = self.region {
+                // Discard the subdivison if it doesn't match the region
+                (
+                    Some(r),
+                    Some(*sd).filter(|sd| sd.region == r && !sd.suffix.is_unknown()),
+                )
+            } else {
+                // Use the subdivision's region if there's no region
+                (
+                    Some(sd.region),
+                    Some(*sd).filter(|sd| !sd.suffix.is_unknown()),
+                )
+            }
+        } else {
+            (self.region, None)
+        };
+
         DataLocale {
             language: self.language,
             script: self.script,
-            region: match (self.region, self.region_override) {
-                (_, Some(r)) if region_priority => Some(r.region),
-                (r, _) => r,
-            },
+            region,
             variant: self.variant,
-            subdivision: self.subdivision.map(|r| r.into_subtag()),
+            subdivision: subdivision.map(|r| r.into_subtag()),
         }
     }
 
-    /// Convert to a DataLocale, with region-based fallback priority
-    ///
-    /// Most users should use `icu_provider::marker::make_locale()` instead.
-    pub fn to_data_locale_region_priority(self) -> DataLocale {
-        self.to_data_locale_maybe_region_priority(true)
-    }
-
     /// Convert to a DataLocale, with language-based fallback priority
     ///
     /// Most users should use `icu_provider::marker::make_locale()` instead.
     pub fn to_data_locale_language_priority(self) -> DataLocale {
-        self.to_data_locale_maybe_region_priority(false)
+        let subdivision = self.subdivision.filter(|sd| {
+            // Discard the subdivision if it doesn't match the region
+            self.region
+                .is_none_or(|r| r == sd.region && !sd.suffix.is_unknown())
+        });
```

Comment:
There's no harm in making en-u-sd-ustx map to en-US-u-sd-ustx, since the fallback chain will still reach it, so I think we should do that and then DataLocale can store just the subtag suffix in that field.

Actually either one of those two locales is okay, but I'd rather pick one as the only form that we ever evaluate in fallback and DataLocale.